### ISSUE TITLE
NVBug 2318871: Compilation failure with `zip_iterator` and `complex<float>` in CUDA 9.2

### ISSUE THREAD

**Issue by pijyoi** [thrust/thrust#919]
CUDA version: 9.2
Thrust version: 1.9.2 (bundled in CUDA 9.2)
GCC version: 5.4.0

The following code fails to compile for cuda backend, specifically the line "auto zipout_end = ...".
It compiles however if we make any of the following changes:
1) change TYPE to float
2) target openmp backend.

```c++
#include <thrust/device_vector.h>
#include <thrust/sequence.h>
#include <thrust/copy.h>
#include <thrust/gather.h>
#include <thrust/iterator/counting_iterator.h>
#include <thrust/iterator/zip_iterator.h>
#include <thrust/complex.h>
 
/* g++ -std=c++11 -I/usr/local/cuda/include -O2 -x c++ -fopenmp -DTHRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_OMP -lgomp minimal.cu
 nvcc -std=c++11 --expt-extended-lambda minimal.cu
*/
 
// typedef float TYPE;
typedef thrust::complex<float> TYPE;
 
int main()
{
   thrust::device_vector<TYPE> d_vec(10);
   thrust::sequence(d_vec.begin(), d_vec.end());
   thrust::device_vector<TYPE> d_res(10);
 
   auto pred = [] __host__ __device__ (TYPE val) { return abs(val) > 5; };
   auto cntit_begin = thrust::make_counting_iterator(0);
   auto cntit_end = cntit_begin + 10;
 
   // generate indices and values in two calls
   thrust::device_vector<int> indices(10);
   auto indices_end = thrust::copy_if(cntit_begin, cntit_end, d_vec.begin(), indices.begin(), pred);
   thrust::gather(indices.begin(), indices_end, d_vec.begin(), d_res.begin());
 
   // generate indices and values in one call
   auto zipin_begin = thrust::make_zip_iterator(thrust::make_tuple(cntit_begin, d_vec.begin()));
   auto zipin_end = thrust::make_zip_iterator(thrust::make_tuple(cntit_end, d_vec.end()));
   auto zipout_begin = thrust::make_zip_iterator(thrust::make_tuple(indices.begin(), d_res.begin()));
   // the following line fails to compile for combination of cuda backend and complex type
   auto zipout_end = thrust::copy_if(zipin_begin, zipin_end, d_vec.begin(), zipout_begin, pred);
}
```

**Comment by brycelelbach** [comment]
Right now we have a testing gap internally - we don't really have test coverage for the OpenMP backend. I'm working on changing that. We do have an open TODO item for repairing support for the OpenMP backend.

**Comment by GregDMeyer** [comment]
Probably related:

The following extremely simple example fails to compile (the corresponding bug prevents building the PETSc library with CUDA 9.2):

```
#include <thrust/device_vector.h>                                                                                                    
#include <thrust/complex.h>                                                                                                          
                                                                                                                                     
void test_function() {                                                                                                               
                                                                                                                                     
  thrust::device_vector< thrust::complex<double> > v;                                                                                
  v.resize(1);                                                                                                                       
                                                                                                                                     
}
```

**Comment by brycelelbach** [comment]
@GregDMeyer, that's with the OpenMP backend?

**Comment by FilipeMaia** [comment]
@brycelelbach, the example from @GregDMeyer also fails with the CUDA backend:

```
$ nvcc -c complex.cu
/usr/local/cuda-9.2/bin/../targets/x86_64-linux/include/thrust/detail/complex/complex.inl(187): error: no instance of overloaded function "thrust::complex<T>::real [with T=double]" matches the argument list
            argument types are: (const thrust::reference<thrust::complex<double>, thrust::pointer<thrust::complex<double>, thrust::cuda_cub::tag, thrust::use_default, thrust::use_default>, thrust::use_default>)
          detected during:
            instantiation of "thrust::complex<T> &thrust::complex<T>::operator=(const R &) [with T=double, R=thrust::reference<thrust::complex<double>, thrust::pointer<thrust::complex<double>, thrust::cuda_cub::tag, thrust::use_default, thrust::use_default>, thrust::use_default>]"
/usr/local/cuda-9.2/bin/../targets/x86_64-linux/include/thrust/system/detail/sequential/general_copy.h(76): here
            instantiation of "thrust::detail::enable_if<thrust::system::detail::sequential::general_copy_detail::reference_is_assignable<InputIterator, OutputIterator>::value, void>::type thrust::system::detail::sequential::general_copy_detail::iter_assign(OutputIterator, InputIterator) [with OutputIterator=thrust::complex<double> *, InputIterator=thrust::pointer<thrust::complex<double>, thrust::cuda_cub::tag, thrust::use_default, thrust::use_default>]"
/usr/local/cuda-9.2/bin/../targets/x86_64-linux/include/thrust/system/detail/sequential/general_copy.h(112): here
            instantiation of "OutputIterator thrust::system::detail::sequential::general_copy(InputIterator, InputIterator, OutputIterator) [with InputIterator=thrust::pointer<thrust::complex<double>, thrust::cuda_cub::tag, thrust::use_default, thrust::use_default>, OutputIterator=thrust::complex<double> *]"
/usr/local/cuda-9.2/bin/../targets/x86_64-linux/include/thrust/system/detail/sequential/copy.inl(74): here
            instantiation of "OutputIterator thrust::system::detail::sequential::copy_detail::copy(InputIterator, InputIterator, OutputIterator, thrust::detail::false_type) [with InputIterator=thrust::pointer<thrust::complex<double>, thrust::cuda_cub::tag, thrust::use_default, thrust::use_default>, OutputIterator=thrust::complex<double> *]"
/usr/local/cuda-9.2/bin/../targets/x86_64-linux/include/thrust/system/detail/sequential/copy.inl(121): here
            instantiation of "OutputIterator thrust::system::detail::sequential::copy(thrust::system::detail::sequential::execution_policy<DerivedPolicy> &, InputIterator, InputIterator, OutputIterator) [with DerivedPolicy=thrust::detail::seq_t, InputIterator=thrust::pointer<thrust::complex<double>, thrust::cuda_cub::tag, thrust::use_default, thrust::use_default>, OutputIterator=thrust::complex<double> *]"
/usr/local/cuda-9.2/bin/../targets/x86_64-linux/include/thrust/detail/copy.inl(37): here
            [ 15 instantiation contexts not shown ]
            instantiation of "OutputIt thrust::copy(const thrust::detail::execution_policy_base<DerivedPolicy> &, InputIt, InputIt, OutputIt) [with DerivedPolicy=thrust::cuda_cub::tag, InputIt=thrust::detail::normal_iterator<thrust::pointer<thrust::complex<double>, thrust::cuda_cub::tag, thrust::use_default, thrust::use_default>>, OutputIt=thrust::detail::normal_iterator<thrust::device_ptr<thrust::complex<double>>>]"
/usr/local/cuda-9.2/bin/../targets/x86_64-linux/include/thrust/detail/overlapped_copy.h(106): here
            instantiation of "RandomAccessIterator2 thrust::detail::dispatch::overlapped_copy(thrust::execution_policy<DerivedPolicy> &, RandomAccessIterator1, RandomAccessIterator1, RandomAccessIterator2) [with DerivedPolicy=thrust::cuda_cub::tag, RandomAccessIterator1=thrust::detail::normal_iterator<thrust::device_ptr<thrust::complex<double>>>, RandomAccessIterator2=thrust::detail::normal_iterator<thrust::device_ptr<thrust::complex<double>>>]"
/usr/local/cuda-9.2/bin/../targets/x86_64-linux/include/thrust/detail/overlapped_copy.h(126): here
            instantiation of "RandomAccessIterator2 thrust::detail::overlapped_copy(RandomAccessIterator1, RandomAccessIterator1, RandomAccessIterator2) [with RandomAccessIterator1=thrust::detail::normal_iterator<thrust::device_ptr<thrust::complex<double>>>, RandomAccessIterator2=thrust::detail::normal_iterator<thrust::device_ptr<thrust::complex<double>>>]"
/usr/local/cuda-9.2/bin/../targets/x86_64-linux/include/thrust/detail/vector_base.inl(546): here
            instantiation of "thrust::detail::vector_base<T, Alloc>::iterator thrust::detail::vector_base<T, Alloc>::erase(thrust::detail::vector_base<T, Alloc>::iterator, thrust::detail::vector_base<T, Alloc>::iterator) [with T=thrust::complex<double>, Alloc=thrust::device_malloc_allocator<thrust::complex<double>>]"
/usr/local/cuda-9.2/bin/../targets/x86_64-linux/include/thrust/detail/vector_base.inl(257): here
            instantiation of "void thrust::detail::vector_base<T, Alloc>::resize(thrust::detail::vector_base<T, Alloc>::size_type) [with T=thrust::complex<double>, Alloc=thrust::device_malloc_allocator<thrust::complex<double>>]"
complex.cu(7): here

1 error detected in the compilation of "/tmp/tmpxft_000051e5_00000000-6_complex.cpp1.ii".
```

**Comment by GregDMeyer** [comment]
Thanks @FilipeMaia!

@brycelelbach I didn't try with the openMP backend, only with CUDA. But if you look at the original question--it's actually referencing the CUDA backend (though a bit ambiguous):

> The following code fails to compile for cuda backend

**Comment by pijyoi** [comment]
Yes it's the cuda backend for which I had an issue.
I didn't realise that it had been misunderstood that I had an issue with the openmp backend.

**Comment by FilipeMaia** [comment]
Here's another, hopefully simpler, function reproducing the same problem:
```c++
#include <thrust/device_ptr.h>
#include <thrust/complex.h>

void test_function() {
  thrust::complex<double> a;
  thrust::device_ptr<thrust::complex<double> > p;
  a = *p;
}
```

```bash
$ nvcc -c complex3.cu
/usr/local/cuda-9.2/bin/../targets/x86_64-linux/include/thrust/detail/complex/complex.inl(187): error: no instance of overloaded function "thrust::complex<T>::real [with T=double]" matches the argument list
            argument types are: (const thrust::device_reference<thrust::complex<double>>)
          detected during instantiation of "thrust::complex<T> &thrust::complex<T>::operator=(const R &) [with T=double, R=thrust::device_reference<thrust::complex<double>>]"
complex3.cu(7): here

1 error detected in the compilation of "/tmp/tmpxft_000044a6_00000000-6_complex3.cpp1.ii".
```

**Comment by jaredhoberock** [comment]
I would have expected `thrust::reference<T>` or `thrust::device_reference<T>` to have a conversion-to-T operator that would handle this case. Doesn't it? I don't know why `thrust::complex` would cause problems.

**Comment by jaredhoberock** [comment]
Maybe the issue is that `thrust::complex<T>::operator=(const R&)` is a better match than the conversion operator. `std::complex` does not have a templated `operator=`. This could be the issue.

**Comment by FilipeMaia** [comment]
I think @jaredhoberock is right. Removing `thrust::complex<T>::operator=(const R&)` fixes the problem.

**Comment by brycelelbach** [comment]
GH #928 / NVBug 2341455 is probably a duplicate of this.

**Comment by griwes** [comment]
@pijyoi Can you please check if what you are trying to do works on the version of Thrust shipped with CUDA 10.0?

**Comment by pijyoi** [comment]
@griwes I have just downloaded CUDA 10.0 and the code snippet included in this issue compiles successfully.

**Comment by brycelelbach** [comment]
Fixed as of CUDA 10.0

**Comment by FilipeMaia** [comment]
I suggest you make a comment pointing to this issue instead of just wrapping the relevant code in `#if 0 #endif`, as it's not entirely obvious what's the problem just by looking at it.
